examples/misc/packed_bubbles.py

```diff
@@ -0,0 +1,162 @@
+            rest_bub = np.delete(self.bubbles, i, 0)
+            # try to move directly towoards the center of mass
+            # dir_vec from bubble to com
+            dir_vec = self.com - self.bubbles[i, :2]
+
+            # shorten dir_vec to have length of 1
+            dir_vec = dir_vec / np.sqrt(dir_vec.dot(dir_vec))
+
+            # calculate new bubble position
+            new_point = self.bubbles[i, :2] + dir_vec * self.step_dist
+            new_bubble = np.append(new_point, self.bubbles[i, 2:4])
+
+            # check whether new bubble collides with oder bubbles
+            if not self.check_collisions(new_bubble, rest_bub):
+                self.bubbles[i, :] = new_bubble
+                self.com = self.center_of_mass()
+                moves += 1
+            else:
+                # try to move around a bubble that you collide with
+                # find colliding bubble
+                for colliding in self.collides_with(new_bubble, rest_bub):
+                    # calculate dir vec
+                    dir_vec = rest_bub[colliding, :2] - self.bubbles[i, :2]
+                    dir_vec = dir_vec / np.sqrt(dir_vec.dot(dir_vec))
+                    # calculate orthagonal vec
+                    orth = np.array([dir_vec[1], -dir_vec[0]])
+                    # test which dir to go
+                    new_point1 = self.bubbles[i, :2] + orth * self.step_dist
+                    new_point2 = self.bubbles[i, :2] - orth * self.step_dist
+                    dist1 = self.center_distance(
+                        self.com, np.array([new_point1]))
+                    dist2 = self.center_distance(
+                        self.com, np.array([new_point2]))
+                    new_point = new_point1 if dist1 < dist2 else new_point2
+                    new_bubble = np.append(new_point, self.bubbles[i, 2:4])
+                    if not self.check_collisions(new_bubble, rest_bub):
+                        self.bubbles[i, :] = new_bubble
+                        self.com = self.center_of_mass()
+
+        if moves / len(self) < 0.1:
+            self.step_dist = self.step_dist / 2
+
+    def plot(self, ax, labels, colors):
+        """
+        draw the bubble plot
+
+        Args:
+            ax (matplotlib.axes._subplots.AxesSubplot)
+            labels (list): labels of the bubbles
+            colors (list): colors of the bubbles
+        """
+        for i in range(len(self)):
+            circ = plt.Circle(
+                self.bubbles[i, :2], self.bubbles[i, 2], color=colors[i])
+            ax.add_patch(circ)
+            ax.text(*self.bubbles[i, :2], labels[i],
+                    horizontalalignment='center', verticalalignment='center')
+
+
+# set market share of the browsers as area of the bubbles
+bubble_plot = BubbleChart(a=np.array(
+    browser_market_share['market_share']), bubble_distance=1)
+
+# collapse plot 50 times. In some cases it might be useful
+# to do this more or less often
+for i in range(50):
```

Comment:
I'm not quite following the loop here.  What happens if you only do this once versus 50 times?

Reply:
Each loop further packs the plot by moving the bubbles. (the bubbles move less each time fewer than 10% of bubbles could move without colliding.) The following pictures show the result when running the loop 0, 1, 10 or 50 times. When using 3000 bubbles, you might only want to run the loop 20 times, when you want to perfectly pack 4 Bubbles, you might run it 300 times.
### 0 loops
![0](https://user-images.githubusercontent.com/36071676/90042789-6d648f00-dccb-11ea-9c41-7392652c33e6.png)
### 1 loop
![1](https://user-images.githubusercontent.com/36071676/90042787-6ccbf880-dccb-11ea-80c1-391cb7a5cf2f.png)
### 10 loops
![10](https://user-images.githubusercontent.com/36071676/90042783-6c336200-dccb-11ea-9c64-8d99eeaec5ae.png)
### 50 loops
![50](https://user-images.githubusercontent.com/36071676/90042780-6b023500-dccb-11ea-9129-3c69c80f4ced.png)

Reply:
This seems closely related to the algorithms behind kmeans. Is there anything to be borrowed from that

Reply:
I came up with everything myself, but it can still have some similarities.